<a href="https://colab.research.google.com/github/DiogoSabec/Bootstrap-test/blob/main/Colabs/AI/Perceptron_MLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Configuração

Import do dataset da Iris e do MLP, utilizando o sklearn

In [ ]:
import numpy as np

from sklearn.datasets import load_iris
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split

Carregamento da Iris

In [ ]:
iris = load_iris()

arrays = iris.data

Separação dos inputs e dos outputs (x e y)

In [ ]:
X = arrays[:,:]
X

data = load_iris()
y = data.target
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

*   Utilize como divisão de treino e teste a configuração de 70% por 30%. ✅
*   Garanta o embaralhamento dos dados antes de submetê-los ao treinamento da rede neural. ✅ (O train_test_split() embaralha os dados)



In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Combinações com camadas ocultas

## Todas as combinações com 1 camada oculta, de 1 a 100

In [ ]:
best_score = 0
best_params = None

for neurons_layer1 in range(1, 101):
      clf = MLPClassifier(
          hidden_layer_sizes=[neurons_layer1],
          activation='logistic',
          max_iter= 5000,
          random_state=42,
          early_stopping=True,
      )

      clf.fit(X_train, y_train)
      score = clf.score(X_test, y_test)

      print(f"{neurons_layer1}, Score: {score}")

      if score > best_score and score != 1.0 and score:
          best_score = score
          best_params = (neurons_layer1)

print("\nBest Parameters:")
print(f"{best_params}, Score: {best_score}")


1, Score: 0.4444444444444444
2, Score: 0.28888888888888886
3, Score: 0.28888888888888886
4, Score: 0.28888888888888886
5, Score: 0.28888888888888886
6, Score: 0.28888888888888886
7, Score: 0.4222222222222222
8, Score: 0.28888888888888886
9, Score: 0.28888888888888886
10, Score: 0.17777777777777778
11, Score: 0.4222222222222222
12, Score: 0.28888888888888886
13, Score: 0.28888888888888886
14, Score: 0.2222222222222222
15, Score: 0.28888888888888886
16, Score: 0.28888888888888886
17, Score: 0.26666666666666666
18, Score: 0.28888888888888886
19, Score: 0.28888888888888886
20, Score: 0.28888888888888886
21, Score: 0.4666666666666667
22, Score: 0.28888888888888886
23, Score: 0.24444444444444444
24, Score: 0.28888888888888886
25, Score: 0.28888888888888886
26, Score: 0.15555555555555556
27, Score: 0.4444444444444444
28, Score: 0.28888888888888886
29, Score: 0.4222222222222222
30, Score: 0.28888888888888886
31, Score: 0.28888888888888886
32, Score: 0.7111111111111111
33, Score: 0.288888888888

**Aqui podemos ver que com 32 neuronios em uma única camada oculta, consegue atingir um valor razoável e melhor que quase 70 dos neuronios restantes, mostrando que nem sempre a quantidade ajuda.**

In [ ]:
best_score = 0
best_params = None

for neurons_layer1 in range(1, 101):
  clf = MLPClassifier(
    hidden_layer_sizes=[neurons_layer1],
    activation='logistic',
    max_iter= 5000,
    alpha=0.001,
    batch_size=50,
    learning_rate_init=0.01,
    random_state=0,
    early_stopping=True,
      )

  clf.fit(X_train, y_train)
  score = clf.score(X_test, y_test)

  print(f"{neurons_layer1}, Score: {score}")

  if score > best_score and score != 1.0 and score:
      best_score = score
      best_params = (neurons_layer1)

print("\nBest Parameters:")
print(f"{best_params}, Score: {best_score}")


1, Score: 0.6444444444444445
2, Score: 0.4
3, Score: 0.5111111111111111
4, Score: 0.28888888888888886
5, Score: 0.9333333333333333
6, Score: 0.5777777777777777
7, Score: 0.7333333333333333
8, Score: 0.9555555555555556
9, Score: 0.28888888888888886
10, Score: 0.7111111111111111
11, Score: 0.4666666666666667
12, Score: 0.8222222222222222
13, Score: 0.7555555555555555
14, Score: 0.5333333333333333
15, Score: 0.8222222222222222
16, Score: 0.9777777777777777
17, Score: 0.9555555555555556
18, Score: 0.8666666666666667
19, Score: 0.8666666666666667
20, Score: 0.8666666666666667
21, Score: 0.9777777777777777
22, Score: 0.9777777777777777
23, Score: 0.9777777777777777
24, Score: 0.9777777777777777
25, Score: 0.7777777777777778
26, Score: 0.7777777777777778
27, Score: 0.8444444444444444
28, Score: 0.8
29, Score: 0.8222222222222222
30, Score: 0.7777777777777778
31, Score: 0.8222222222222222
32, Score: 0.9777777777777777
33, Score: 0.9777777777777777
34, Score: 0.9111111111111111
35, Score: 0.9111

**Com os parâmetros diferentes, conseguimos o um resultado melhor com uma quantidade menor de neurônios, sendo mais eficiente**

##Todas as combinações com 2 camadas ocultas, de 1 a 100

In [ ]:
best_score = 0
best_params = None

for neurons_layer1 in range(1, 101):
    for neurons_layer2 in range(1, 101):
        clf = MLPClassifier(
            hidden_layer_sizes=[neurons_layer1, neurons_layer2],
            activation='logistic',
            max_iter= 5000,
            random_state=42,
            early_stopping=True,
        )

        clf.fit(X_train, y_train)
        score = clf.score(X_test, y_test)

        print(f"{neurons_layer1} : {neurons_layer2}, Score: {score}")

        if score > best_score and score != 1.0:
            best_score = score
            best_params = (neurons_layer1, neurons_layer2)

print("\nBest Parameters:")
print(f"{best_params[0]} : {best_params[1]}, Score: {best_score}")


A saída de streaming foi truncada nas últimas 5000 linhas.
51 : 4, Score: 0.4222222222222222
51 : 5, Score: 0.28888888888888886
51 : 6, Score: 0.28888888888888886
51 : 7, Score: 0.28888888888888886
51 : 8, Score: 0.4222222222222222
51 : 9, Score: 0.28888888888888886
51 : 10, Score: 0.28888888888888886
51 : 11, Score: 0.7111111111111111
51 : 12, Score: 0.28888888888888886
51 : 13, Score: 0.28888888888888886
51 : 14, Score: 0.28888888888888886
51 : 15, Score: 0.4222222222222222
51 : 16, Score: 0.28888888888888886
51 : 17, Score: 0.28888888888888886
51 : 18, Score: 0.28888888888888886
51 : 19, Score: 0.28888888888888886
51 : 20, Score: 0.28888888888888886
51 : 21, Score: 0.4666666666666667
51 : 22, Score: 0.28888888888888886
51 : 23, Score: 0.28888888888888886
51 : 24, Score: 0.28888888888888886
51 : 25, Score: 0.28888888888888886
51 : 26, Score: 0.4222222222222222
51 : 27, Score: 0.28888888888888886
51 : 28, Score: 0.28888888888888886
51 : 29, Score: 0.28888888888888886
51 : 30, Score: 0

**Esse códgo demora a sua execução por quanta da quantidade de combinações, porém podemos ver que a IA não precisa de mais de 50 neuronios em sua camada, indicando que não precisa de muito para que ela seja eficiente.**

In [ ]:
best_score = 0
best_params = None

for neurons_layer1 in range(1, 21):
    for neurons_layer2 in range(1, 21):
        clf = MLPClassifier(
            hidden_layer_sizes=[neurons_layer1, neurons_layer2],
            activation='logistic',
            max_iter= 5000,
            alpha=0.001,
            batch_size=50,
            learning_rate_init=0.01,
            random_state=0,
            early_stopping=True,
        )

        clf.fit(X_train, y_train)
        score = clf.score(X_test, y_test)

        print(f"{neurons_layer1} : {neurons_layer2}, Score: {score}")

        if score > best_score and score != 1.0:
            best_score = score
            best_params = (neurons_layer1, neurons_layer2)

print("\nBest Parameters:")
print(f"{best_params[0]} : {best_params[1]}, Score: {best_score}")


1 : 1, Score: 0.4222222222222222
1 : 2, Score: 0.4666666666666667
1 : 3, Score: 0.4222222222222222
1 : 4, Score: 0.28888888888888886
1 : 5, Score: 0.28888888888888886
1 : 6, Score: 0.4222222222222222
1 : 7, Score: 0.28888888888888886
1 : 8, Score: 0.28888888888888886
1 : 9, Score: 0.28888888888888886
1 : 10, Score: 0.28888888888888886
1 : 11, Score: 0.28888888888888886
1 : 12, Score: 0.28888888888888886
1 : 13, Score: 0.28888888888888886
1 : 14, Score: 0.28888888888888886
1 : 15, Score: 0.4222222222222222
1 : 16, Score: 0.28888888888888886
1 : 17, Score: 0.28888888888888886
1 : 18, Score: 0.28888888888888886
1 : 19, Score: 0.28888888888888886
1 : 20, Score: 0.28888888888888886
2 : 1, Score: 0.28888888888888886
2 : 2, Score: 0.28888888888888886
2 : 3, Score: 0.28888888888888886
2 : 4, Score: 0.28888888888888886
2 : 5, Score: 0.28888888888888886
2 : 6, Score: 0.4
2 : 7, Score: 0.28888888888888886
2 : 8, Score: 0.28888888888888886
2 : 9, Score: 0.28888888888888886
2 : 10, Score: 0.2888888

**Ao rodar esse código com 20:20 e algumas alterções, é possível ver que atingimos um resultado próximo ao anterior, com uma quantidade muito menor de neuronios em cada camada**

## Todas as combinações com 3 camadas ocultas, de 1 a 20

In [ ]:
best_score = 0
best_params = None

for neurons_layer1 in range(1, 21):
    for neurons_layer2 in range(1, 21):
      for neurons_layer3 in range(1, 21):
        clf = MLPClassifier(
            hidden_layer_sizes=[neurons_layer1, neurons_layer2, neurons_layer3],
            activation='logistic',
            max_iter= 5000,
            random_state=42,
            early_stopping=True,
        )

        clf.fit(X_train, y_train)
        score = clf.score(X_test, y_test)

        print(f"{neurons_layer1} : {neurons_layer2}: {neurons_layer3}, Score: {score}")

        if score > best_score and score != 1.0 and score:
            best_score = score
            best_params = (neurons_layer1, neurons_layer2, neurons_layer3)

print("\nBest Parameters:")
print(f"{best_params[0]} : {best_params[1]} : {best_params[2]}, Score: {best_score}")


A saída de streaming foi truncada nas últimas 5000 linhas.
8 : 11: 4, Score: 0.28888888888888886
8 : 11: 5, Score: 0.28888888888888886
8 : 11: 6, Score: 0.4222222222222222
8 : 11: 7, Score: 0.28888888888888886
8 : 11: 8, Score: 0.28888888888888886
8 : 11: 9, Score: 0.28888888888888886
8 : 11: 10, Score: 0.28888888888888886
8 : 11: 11, Score: 0.4222222222222222
8 : 11: 12, Score: 0.4222222222222222
8 : 11: 13, Score: 0.4222222222222222
8 : 11: 14, Score: 0.4222222222222222
8 : 11: 15, Score: 0.4222222222222222
8 : 11: 16, Score: 0.4222222222222222
8 : 11: 17, Score: 0.4222222222222222
8 : 11: 18, Score: 0.28888888888888886
8 : 11: 19, Score: 0.28888888888888886
8 : 11: 20, Score: 0.28888888888888886
8 : 12: 1, Score: 0.28888888888888886
8 : 12: 2, Score: 0.28888888888888886
8 : 12: 3, Score: 0.28888888888888886
8 : 12: 4, Score: 0.4222222222222222
8 : 12: 5, Score: 0.28888888888888886
8 : 12: 6, Score: 0.28888888888888886
8 : 12: 7, Score: 0.28888888888888886
8 : 12: 8, Score: 0.4222222

**Foi utilizado uma quantidade maior de camadas, com menos quantidade de neurônios, podemos ver que utilizando uma baixa quantidade, também chegamos a um resultado próximo a quantidade de valores utilizando apenas 1 camada, que foi 32 neurônios**

# Resultados:


1 Camada Oculta
*   Melhor resultado entre 1 e 100 neuronios: **32, Score: 0.7111111111111111**
*   Melhor resultado entre 1 e 100 neuronios com diferentes parâmetros: **16, Score: 0.9777777777777777**

2 Camadas Ocultas
*   Melhor resultado entre 1 e 100 neuronios: **20 : 58, Score: 0.9777777777777777**
*   Melhor resultado entre 1 e 100 neuronios com diferentes parâmetros: **3 : 7, Score: 0.9777777777777777**

3 Camadas Ocultas
*   Melhor resultado entre 1 e 100 neuronios: **2 : 1 : 8, Score: 0.7111111111111111**



Comparando os resultados, o valor mais otimizado é o com duas camadas, com parâmetros diferentes, pois utiliza um numero menor de neurônios que os demais.